### PR TITLE
PTFE-584 bootstrap project design and structure

### PR DESCRIPTION
**Pull request by tcarmet**
Initial design for upcoming 1.0 project's release.

### CODE REVIEW

File: docs/development/configuration.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+- The GitHub Authentication parameters. (Required)
+  - For GitHub Application: The GitHub Application ID, Installation ID and Private Key. (Required)
+  - For GitHub Personal Access Token: The GitHub Personal Access Token. (Required)
+- The Redis connection parameters. (Required)
+- The backends configuration. (Required)
+- The webhook secret. (Required)
+- The runner groups. (Required)
+- The health check interval. (Default: 10 minute)
+- The runner's time to start. (Default: 10 minute)
+- The runner's time to live. (Default: 12 hour)
```
Comment (gaspardmoindrot) [inline docs/development/configuration.md:26]:
```suggestion
- The health check interval. (Default: 10 minutes)
- The runner's time to start. (Default: 10 minutes)
- The runner's time to live. (Default: 12 hours)
```

File: docs/development/configuration.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+- Workflow access: a comma separated list of the workflows that can access the runner group.
+  For example:
+  ```shell
+  octo-org/octo-repo/.github/workflows/build.yml@v2,
+  octo-org/octo-repo/.github/workflows/deploy.yml@d6dc6c96df4f32fa27b039f2084f576ed2c5c2a5,
+  monalisa/octo-test/.github/workflows/test.yml@main
+  ```
+- The name of the workflow that will be used to run the jobs. (Optional)
+- The maximum number of runners that can run simultaneously. (Default: 20)
+- The minimum number of runners that must be available. (To be reviewed) (Default: 0)
```
Comment (gaspardmoindrot) [inline docs/development/configuration.md:49]: What does "(To be reviewed)" mean ? It's not implemented yet and till you don't implement it you'll let this "to be reviewed" ?

File: docs/development/environment.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# Developer Environment
+
+A devcontainer is provided to ease the development of the runner manager.
+It can be used with GitHub's Codespaces or Visual Studio Code.
+
+It is to be considered as the only supported development environment
+for ease of support and reproducibility.
```
Comment (gaspardmoindrot) [inline docs/development/environment.md:7]: There will be no chart or kustomize ?

File: docs/development/index.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+Information needed to develop and contribute to the project.
+
+## Table of Contents
+
+- [Concepts](./concepts.md)
+- [Workflows](./workflows.md)
+- [Configuration](./configuration.md)
+- [Code](./code.md)
+- [Testing](./testing.md)
+- [Local environment](./environment.md)
```
Comment (gaspardmoindrot) [inline docs/development/index.md:12]:
```suggestion
- [Concepts](./concepts.md).
- [Workflows](./workflows.md).
- [Configuration](./configuration.md).
- [Code](./code.md).
- [Testing](./testing.md).
- [Local environment](./environment.md).
```

File: docs/development/testing.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
+# Testing
+
+Tests will be written using [pytest](https://docs.pytest.org/en/stable/).
+
+The following type of tests wil be written:
+
+- Unit tests
+- API tests
+- Functional tests
```
Comment (gaspardmoindrot) [inline docs/development/testing.md:9]:
```suggestion
- Unit tests.
- API tests.
- Functional tests.
```

File: docs/development/testing.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+- A real rq worker to process the jobs.
+- A mock backend, no real runners will be created.
+- [GitHub's API mock](#githubs-api).
+
+## Functional tests
+
+The functional tests of the runner-manager will be done with no mocking, it will:
+
+- Interact with the real GitHub API.
+- Execute a [GitHub Actions] a real workflow and triggered by a workflow dispatch
```
Comment (gaspardmoindrot) [inline docs/development/testing.md:86]:
```suggestion
- Execute a real workflow using [GitHub Actions] triggered by a workflow dispatch.
```

File: docs/development/testing.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+- Receive webhooks notification from github, thanks to the integration of webhook redirection in `gh` cli.
```
Comment (gaspardmoindrot) [inline docs/development/testing.md:87]: Not sure
```suggestion
- Receive webhooks notification from GitHub, thanks to the integration of webhook redirection in `gh` cli.
```

File: docs/development/workflows.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+A high level description of the runner manager workflows.
+
+## Webhook events
+
+Webhook events are sent by GitHub on selected events, such as:
+
+- Push to a repository
+- Pull requests is merged
+- A workflow is triggered
```
Comment (gaspardmoindrot) [inline docs/development/workflows.md:11]:
```suggestion
- Push to a repository.
- Pull requests is merged.
- A workflow is triggered.
```

File: docs/development/workflows.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        else The maximum number of runners is reached
+            Runner Manager ->> Runner Manager: Try again later
+        end
+    else The job status is running
+        Runner Manager ->> Backend: Update the runner status and metadata
+    else The job status is completed
+        Runner Manager ->> Backend: Delete the runner
+        Backend ->> Runner: Deletes the runner
+        deactivate Runner
+    end
```
Comment (gaspardmoindrot) [inline docs/development/workflows.md:41]: Really really good

File: runner_manager/models/runner.py [hunk 10]
```diff
@@ -1,6 +1,35 @@
...
+    online = "online"
+    idle = "idle"
+    offline = "offline"
+
+
+class Runner(BaseModel):
+    name: str = Field(index=True)
+    runner_group_id: int = Field(ge=0, index=True)
+    instance_id: Optional[int] = Field(ge=0, index=True)
+    status: str = "offline"
```
Comment (gaspardmoindrot) [inline runner_manager/models/runner.py:29]: Shouldn't you use the RunnerStatus class here precisely?

File: tests/images/github-mock/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,9 @@
+FROM ghcr.io/stoplightio/prism/prism:master
+
+HEALTHCHECK NONE
+
+USER node
+
+EXPOSE 4010
+
+CMD ["mock", "https://raw.githubusercontent.com/github/rest-api-description/main/descriptions/ghec/ghec.2022-11-28.json", "-h", "0.0.0.0"]
```
Comment (gaspardmoindrot) [inline tests/images/github-mock/Dockerfile:9]:
```suggestion
CMD ["mock", "https://raw.githubusercontent.com/github/rest-api-description/main/descriptions/ghec/ghec.2022-11-28.json", "-h", "0.0.0.0"]

```

**Review by gaspardmoindrot — CHANGES_REQUESTED** [review]
It's amazing the work you've done, it's super clean, especially the documentation. Since it's just the beginning of the work, I don't necessarily have additional things to say as there are still many things to do.

File: docs/development/configuration.md (hunk 2 above)
Comment (tcarmet) [inline docs/development/configuration.md:49]:
```suggestion
- The minimum number of runners that must be available. (Default: 0)
```

File: docs/development/environment.md (hunk 3 above)
Comment (tcarmet) [reply docs/development/environment.md:7]: Not for a developer environment, at least not for now. 

For the service's deployment it's another story.

File: docs/development/workflows.md [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+
+## Webhook events
+
+Webhook events are sent by GitHub on selected events, such as:
+
+- Push to a repository
+- Pull requests is merged
+- A workflow is triggered
+
+In the case of the runner manager, we are interested in `workflow_job` events
```
[the remark this answers is not in the document]
Comment (tcarmet) [reply docs/development/workflows.md:13]: Do we need it? if so, why?

File: runner_manager/models/runner.py (hunk 10 above)
Comment (tcarmet) [reply runner_manager/models/runner.py:29]: Ah yes, you're right, I think I faced an issue with the redis search when using it, I'll have to double check on that.

Comment (tcarmet) [reply runner_manager/models/runner.py]: Got it working after all. Thanks for the catch.

File: docs/development/workflows.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (gaspardmoindrot) [reply docs/development/workflows.md:13]: Indeed, we don't need it since the name of the workflow is specified in workflow_job.

**Review by gaspardmoindrot — APPROVED** [review]
LGTM now, if there are some changes, tell me and I'll do another review

File: docs/development/workflows.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (tcarmet) [reply docs/development/workflows.md:13]: Do we need the name of the workflow as well?

[the remark this answers is not in the document]
Comment (gaspardmoindrot) [reply docs/development/workflows.md:13]: Yes!

**Pull request merged**